### PR TITLE
Encode the visual pattern, closes #233

### LINKED ISSUE #233: Encode the visual verification data as a key
- Parent issues: -
- Relevant issues: -
- Estimated effort: 5

* * *

## User story
As a developer I want to encrypt the public key sent over Bluetooth after using the visual verification activity by the sequence that the user tapped.

## What
This issue demands to encode / decode the public key of the user by the sequence he tapped in the visual verification activity

## Why
This issue is needed for the project because MITM attacks are possible if the information is not encrypted

## How
This feature can be added by starting the key at 0, incrementing by the power of 12 per extra tap and letting each button encode an associated number between 0 and 12.

## Definition of ready
This features is not ready until the following issues are finished:

- 

## Definition of done
This task can be considered done when

- A key is generated based on the sequence of buttons that the user tapped in the visual verification activity
- The key is used to encrypt the public key
- The key is used to decrypt the public key

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: #233 
- Related Pull Requests: -

* * *

## What
This Pull Request adds to the repository an encoding for the pattern tapped in VisualVerificationActivity.
Each button has a value between 1 and 12.
Initially code = 0
Every tap: code = code x 12 + value

## Why
This Pull Request is needed because we need to encode the pattern tapped by the user to encrypt the data send over Bluetooth

## How
This feature can be viewed/tested within the project by... This is only possible after the whole feature is implemented

## Alternative implementation
None considered

## Notes

This PR does not encrypt/decrypt the data. #227 contains a custom send method and a local branch of @kcallebaut contains method to encrypt/decrypt the data. We'll have to wait for those to be merged before we can complete the whole feature

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java [hunk 1]
```diff
@@ -98,16 +100,18 @@ private void nextActivity() {
...
+        if (this.numTaps > VisualVerificationActivity.SECURITY_CODE_LENGTH) {
             LOGGER.warn("Security code already long enough");
-        } else if (this.securityCode.length() + 1 == VisualVerificationActivity.SECURITY_CODE_LENGTH) {
+        } else if (this.numTaps + 1 == VisualVerificationActivity.SECURITY_CODE_LENGTH) {
             this.securityCode += button;
             LOGGER.info("final code is: %s", this.securityCode);
             this.nextActivity();
         } else {
+            this.numTaps++;
+            this.securityCode *= NUM_BUTTONS;
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Why do you multiply it every time with the number of buttons?

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: See the PR description:
> This Pull Request adds to the repository an encoding for the pattern tapped in VisualVerificationActivity.
Each button has a value between 1 and 12.
Initially code = 0
Every tap: code = code x 12 + value

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java [hunk 2]
```diff
@@ -98,16 +100,18 @@ private void nextActivity() {
      * @param v The view of the button being clicked.
      */
     public void buttonAction(final View v) {
-        final String button = v.getContentDescription().toString();
+        final int button = Integer.parseInt(v.getContentDescription().toString());
         LOGGER.info("button '%s' clicked", button);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: I'm not sure how the logger will handle `'%s'` with an int value 🤔

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Not xD
In the logger you should always use {}, that's less error-prone

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: My bad 😅 could you change it tho, there is also a similar log on [line 116](https://github.com/ericcornelissen/NervousFish/pull/250/files/a121f5f95ec29150d5254fccab15f3a42bf7274c#diff-12207c280a7c5d460c92841b36106522R116)

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java [hunk 3]
```diff
@@ -28,9 +28,11 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger("VisualVerificationActivity");
     private static final int SECURITY_CODE_LENGTH = 5;
+    private static final int NUM_BUTTONS = 12;
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Kinda scary to have it hardcoded as our project is open source...
Then again no one will probably use this version of the app 🤔

**Comment by jverbraeken** [comment]
@TheBonheurs This PR doesn't necessarily make encrypting easier, but provides a way to get a certain unique code that always corresponds to the images that were tapped.
This PR is not meant as a huge PR with a thousand lines of code that contains a fully fledged encryption protcol for VisualVerification, but is a small part of the total feature

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java (hunk 3 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: > Kinda scary to have it hardcoded as our project is open source...

Not sure what you mean 😕

Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Everybody knows how many tiles there are on the screen so they can figure out that the picture's animals and objects aren't randomly placed and can more easily seatup for a brute force attack.

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Users can see how many tiles there are anyway. For example, on my emulator I see the button flash when I click on it, very subtle, but I can see that there's a button I'm clicking on and where it is / how large it is

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: > Everybody knows how many tiles there are on the screen so they can figure out that the picture's animals and objects aren't randomly placed and can more easily seatup for a brute force attack.

You mean that they can now find out how many tiles there are on the screen? They already could, the code for that is also part of this repository.

Also, the intention is to have multiple different images for the visual verification method, making already a bit more difficult to do a brute force attack

**Pull request merged**